tests/kernel/timer/timer_behavior/src/main.c

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <math.h>
+#include <stdlib.h>
+
+#include <zephyr/zephyr.h>
+
+#include <tc_util.h>
+#include <ztest.h>
+
+static uint32_t periodic_idx = 0;
+static uint64_t periodic_data[CONFIG_TIMER_TEST_SAMPLES + 1];
+static uint64_t periodic_start, periodic_end;
+static struct k_timer periodic_timer;
+K_SEM_DEFINE(periodic_sem, 0, 1);
+
+void periodic_fn(struct k_timer *t)
+{
+	if (periodic_idx < CONFIG_TIMER_TEST_SAMPLES) {
+
+#ifdef CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+		periodic_end = k_cycle_get_64();
+#else
+		periodic_end = k_cycle_get_32();
+#endif
+		periodic_data[periodic_idx + 1] = periodic_end;
+		periodic_idx++;
+	} else {
+		k_timer_stop(t);
+		k_sem_give(&periodic_sem);
+	}
+}
+
+ZTEST(timer_behavior, periodic_behavior)
+{
+	TC_PRINT("periodic timer behavior test\n");
+
+	k_timer_init(&periodic_timer, periodic_fn, NULL);
+
+#ifdef CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
+	periodic_start = k_cycle_get_64();
+#else
+	periodic_start = k_cycle_get_32();
+#endif
+
+	periodic_data[0] = periodic_start;
+	k_timer_start(&periodic_timer, K_NO_WAIT, K_USEC(CONFIG_TIMER_TEST_PERIOD));
+	k_sem_take(&periodic_sem, K_FOREVER);
+
+	TC_PRINT("periodic timer samples gathered, calculating statistics\n");
+
+	/* calculate variance, and precision */
+	uint64_t total_cycles = 0;
+
+	for (int i = 0; i < CONFIG_TIMER_TEST_SAMPLES; i++) {
+		total_cycles += (periodic_data[i + 1] - periodic_data[i]);
+	}
+
+	double mean = (double)total_cycles / (double)CONFIG_TIMER_TEST_SAMPLES;
+
+	double variance = 0;
+
+	for (int i = 0; i < CONFIG_TIMER_TEST_SAMPLES; i++) {
+		double diff = (double)(periodic_data[i + 1] - periodic_data[i]) - mean;
+		double diff_sq = diff * diff;
+
+		variance += diff_sq;
+	}
+
+	/* A measure of how wide the distribution is, ideal is 0 */
+	variance = variance / (double)CONFIG_TIMER_TEST_SAMPLES;
+
+	/* A measure of timer precision, ideal is 0 */
+	double stddev = sqrtf(variance);
+
+	/* Use double precision math here as integer overflows are possible in doing all the
+	 * conversions otherwise
+	 */
```

Comment:
Disclaimer: I've not delved into the math here to determine how much of an effect (if any) the following is going to have.

I am always wary of floating point in cases of conversion such as this. Yes, they allow for larger numbers to be represented, but at the cost of accuracy.  uint64_t's obviously give us 64 bits to work with. Doubles however only give us 52 bits in the significand (and the implicit 1).